blockrecord/validator_test.go

```diff
@@ -0,0 +1,122 @@
+// license that can be found in the LICENSE file.
+
+package blockrecord_test
+
+import (
+	"testing"
+
+	"github.com/bitmark-inc/bitmarkd/blockdigest"
+	"github.com/bitmark-inc/bitmarkd/blockrecord"
+	"github.com/bitmark-inc/bitmarkd/difficulty"
+	"github.com/bitmark-inc/bitmarkd/fault"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestValidBlockTimeSpacingWhenInitialVersionValid(t *testing.T) {
+	actual := blockrecord.ValidBlockTimeSpacingAtVersion(1, 10)
+	assert.Equal(t, nil, actual, "faile to validate valid v1 blcok time spacing")
+}
+
+func TestValidBlockTimeSpacingWhenCurrentVersionValid(t *testing.T) {
+	actual := blockrecord.ValidBlockTimeSpacingAtVersion(2, uint64(100))
+	assert.Equal(t, nil, actual, "faile to validate valid v2 blcok time spacing")
+}
+
+func TestValidBlockTimeSpacingWhenInitialVersionInvalid(t *testing.T) {
+	actual := blockrecord.ValidBlockTimeSpacingAtVersion(1, uint64(100000))
+	assert.Equal(t, fault.ErrInvalidBlockHeaderTimestamp, actual, "faile to validate invalid v1 blcok time spacing")
+}
+
+func TestValidBlockTimeSpacingWhenCurrentVersionInvalid(t *testing.T) {
+	actual := blockrecord.ValidBlockTimeSpacingAtVersion(2, uint64(10000))
+	assert.Equal(t, fault.ErrInvalidBlockHeaderTimestamp, actual, "faile to validate invalid v2 blcok time spacing")
+}
+
+func TestValidIncomingDifficutyWhenValid(t *testing.T) {
+	difficulty.Current.Set(2)
+	incoming := difficulty.New()
+	incoming.Set(2)
+	actual := blockrecord.ValidIncomingDifficuty(incoming)
+
+	assert.Equal(t, nil, actual, "fail to validate valid incoming difficulty")
+}
+
+func TestValidIncomingDifficutyWhenInValid(t *testing.T) {
+	difficulty.Current.Set(2)
+	incoming := difficulty.New()
+	incoming.Set(4)
+	actual := blockrecord.ValidIncomingDifficuty(incoming)
+
+	assert.Equal(t, fault.ErrDifficultyNotMatch, actual, "fail to validate invalid incoming difficulty")
+}
+
+func TestIsDifficultyAppliedVersionWhenApplied(t *testing.T) {
+	actual := blockrecord.IsDifficultyAppliedVersion(3)
+	assert.Equal(t, true, actual, "fail to check difficulty applied version")
+}
+
+func TestIsDifficultyAppliedVersionWhenNotApplied(t *testing.T) {
+	actual := blockrecord.IsDifficultyAppliedVersion(1)
+	assert.Equal(t, false, actual, "fail to check difficulty not applied version")
+}
+
+func TestValidHeaderVersionWhenTooSmall(t *testing.T) {
+	actual := blockrecord.ValidHeaderVersion(uint16(10), uint16(0))
+	assert.Equal(t, fault.ErrInvalidBlockHeaderVersion, actual, "fail to validate header version too small")
+}
+
+func TestValidHeaderVersionWhenPreviousLarger(t *testing.T) {
+	actual := blockrecord.ValidHeaderVersion(uint16(100), uint16(99))
+	assert.Equal(t, fault.ErrBlockVersionMustNotDecrease, actual, "fail to validate previous version larger")
+}
+
+func TestValidHeaderVersionWhenIncomingLarger(t *testing.T) {
+	actual := blockrecord.ValidHeaderVersion(uint16(100), uint16(101))
+	assert.Equal(t, nil, actual, "fail to validate incoming version larger")
```

Comment:
t.Fatalf

Reply:
It will remove one dependency

Reply:
alternatively if there are other asserts like handling bytes.Equal then we could standardise on assert

Reply:
i write simple test, `assert.Equal` is able to compare bytes.

Reply:
how about adding: `assert := assert.New(t)` then the t parameter is not needed

Reply:
I am ok in both ways, which one do you think better?

1. use `assert.Equal(t, ... )` for only one assertion
    use `assert := assert.New(t), assert.Equal(...)` for more than two assertions in one test

2. use `assert := assert.New(t), assert.Equal(...)` for consistency, regarding how many assertions in a test

Reply:
1 is ok